### PR TITLE
feat: Update types for Cypress.Chainable.task to allow for parameterizing the task return type

### LINKED ISSUE #7136: Typings for cy.task when chaining with .then
### Current behavior:
The function passed to `.then` after calling `.task` must have its parameter typed as `undefined` when writing the code in Typescript. Ideally, the function's parameter would match the return type of the task. As a workaround, the function's parameter can be explicitly marked as `any` then dynamically cast to the desired type, but this goes against Typescript best practices.

### Desired behavior:
There are probably a few different ways of accomplishing this, but one that comes to mind is:
```js
// In this example, val would have type MyType
cy.task<MyType>("task").then((val) => ...);
```

### Test code to reproduce
```js
// In this example, val is typed as undefined
cy.task("task").then((val) => ...);
```

### Versions
Cypress 4.4.0

### PR DESCRIPTION
**Pull request by kevin-wittmer**
Closes #7136

### User facing changelog
Update types for `Cypress.Chainable.task` to allow for parameterizing the task return type.

### Additional details
Previously, the return type of `task` would be determined by the previous subject's type. In this example:

```js
cy.task("task").then((val) => ...);
```

`val` is typed as `undefined` with no way to change it besides explicitly casting `val` inside the lambda function. This change supports specifying the type of `val` in a first-class way:

```js
// val now has type MyType
cy.task<MyType>("task").then((val) => ...);
```

### How has the user experience changed?
User experience has not changed -- developer experience is better!

### PR Tasks
* [x] Have tests been added/updated?
* [x] Has the original issue been tagged with a release in ZenHub? 
* [ ] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
* [ ] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
* [ ] Have new configuration options been added to the [`cypress.schema.json`](https://github.com/cypress-io/cypress/blob/develop/cli/schema/cypress.schema.json)?

### CODE REVIEW

**Review by bahmutov — CHANGES_REQUESTED** [review]
Can you please a few type tests for this?

**Comment by kevin-wittmer** [comment]
@bahmutov Tests have been added. Not sure what type of coverage is expected here or if my approach makes sense, so feedback is welcome. Thanks!

**Review by bahmutov — CHANGES_REQUESTED** [review]
Kevin, could you take a look at these tests failing on CI

https://app.circleci.com/pipelines/github/cypress-io/cypress/13882/workflows/47534892-805e-4efe-b4e0-48e1b93eed78/jobs/473420

```
Installing to /root/.dts/typescript-installs/3.5...
Installed!

Installing to /root/.dts/typescript-installs/next...
Installed!
Error: /root/cypress/cli/types/tests/cypress-tests.ts:608:3
ERROR: 608:3  expect   TypeScript@next expected type to be:
  Cypress.Chainable<unknown>
got:
  Chainable<unknown>
ERROR: 613:3  expect   TypeScript@next expected type to be:
  Cypress.Chainable<number>
got:
  Chainable<number>
ERROR: 617:2  eofline  file should end with a newline

    at /root/cypress/node_modules/dtslint/bin/index.js:190:19
```

**Comment by kevin-wittmer** [comment]
@bahmutov The tests relevant to this PR have been fixed. It doesn't look like the integration test failures are related to this PR, so not sure what's going on there.

File: cli/types/tests/cypress-tests.ts [hunk 1]
```diff
@@ -603,3 +603,10 @@ namespace CypressShadowTests {
   .get('.foo')
   .find('.bar', {includeShadowDom: true})
 }
+
+namespace CypressTaskTests {
+  cy.task<number>('foo') // $ExpectType Chainable<number>
+  cy.task<number>('foo').then((val) => {
```
Comment (bahmutov) [inline cli/types/tests/cypress-tests.ts:609]: and what if the user does not specify a type? Can you add a test
```
cy.task('foo') // $ExpectType Chainable<unknown>
```

Comment (bahmutov) [reply cli/types/tests/cypress-tests.ts:609]: Is the default value yielded from `cy.task` `undefined` or `null`?

**Comment by kevin-wittmer** [comment]
@bahmutov Right, but in your example I would expect the TS code to look like
```
cy.task<null>('foo')
```
because the caller knows that `foo`'s return type is `null`. What I'm getting at is **what do we want the default type to be when the caller does not provide a type?** In concrete terms,
```
cy.task('foo').then(x => ...)
```
what is the type of `x`? Without this PR, the current state is the type of `x` is `undefined`. In the current state of this PR, depending on the version of TS, `x` is either `unknown` or `{}`. I think either `unknown` or `undefined` would be a fine default type. Which do you prefer?

**Comment by bahmutov** [comment]
I vote for unknown

**Comment by jennifer-shehane** [comment]
@kevin-wittmer Do you still have time to address the changes requested in this PR?

**Comment by kevin-wittmer** [comment]
@jennifer-shehane Thanks for the reminder -- I'll address the remaining feedback today.

**Comment by kevin-wittmer** [comment]
@jennifer-shehane @bahmutov Feedback has been addressed.

**Review by flotwig — APPROVED** [review]

**Pull request merged**